docs/design/persistent-storage.md

```diff
@@ -0,0 +1,222 @@
+
+### tl;dr
+
+`PersistentStore` objects contain a VolumeSource with a lifecycle greater than the pods it is used by.  Many types of VolumeSources can be implemented as volume plugins, with each implementing an interface for use within a controller.  A pod with persistence requirements (specifically pod.spec.volumes.volume.persistenceRequirements) is bound to a matching VolumeSource from a pool of available stores.  PersistentStore objects give that VolumeSource longevity.
+
+A `PersistentStoreController` manages the creation, attachment, and reclamation of PersistentStore objects for one VolumeSource.  Implementations must be pluggable to support many types of controllers for many types of VolumeSources. The controller can advise the scheduler on host selection, keeping volume rules contained within the provider.  All cloud provider credentials are isolated to the store controller's host.
+
+`PersistentStoreBinder` watches for pods with persistence requirements and matches the request with what is available in the pool.
+
+`PersistenceRequirements` expresses the capabilities and performance attributes of the underlying VolumeSource.  Allow discovery
+of storage by its capabilities and use those capabilities as pod storage requirements.
+
+Kubernetes makes no guarantees at runtime that the underlying storage exists or is available.  High availability is left to the storage provider.
+
+## Goals
+
+* Allow administrators to describe available storage
+* Allow pod authors to discover and request persistence for their volumes
+* Enforce security through access control lists and securing storage to the same namespace as the pod volume
+* Enforce quotas through admission control
+* Ensure developers can rely on storage being available without being closely bound to a particular disk, server, network, or storage device.
+
+
+#### Allow admins to describe and create available storage
+
+Admins can describe storage using a `PersistentStoreController` for each type of VolumeSource being managed or create storage directly by using the `PersistentStore` API.  Storage created manually still requires a PersistentStoreController and accompanying interface to allow management by the cluster.
+
+
+```
+	
+	PersistentStore
+		TypeMeta
+		ObjectMeta
+			Name - internal name
+			Labels - labeled like Volume.PersistenceRequirements.Selector
+		Spec
+			VolumeSource -- backed by one of these:
+				HostPath
+				EmptyDir
+				GitRepo	
+				GCEPersistentDisk { modes: ReadWriteOnce | ReadOnlyMany }
+				AWSElasticBlockStore { modes: ReadWriteOnce }
+				NFSMount { modes: ReadWriteOnce | ReadOnlyMany | ReadWriteMany }
+				*** many volume plugins
+			// volumes can be orphaned.  allow never delete.
+			RetentionPolicy StoreRetentionPolicy
+		Status
+			Phase
+			CurrentMounts	MountList
+			LastMounts		MountList
+			PodRef			ObjectReference -- the pod originally requesting persistence
+			
+
+	// Manages one type of PersistentStore
+	PersistentStoreController
```

Comment:
I will add more to the doc.

Yes, I intend for the store controller to be an automatic provisioner because it is managing storage of its type.  If it's managing all its volumes, it can advise the scheduler for pod placement, keeping rules isolated to the provider.

One PersistentStoreController per PersistentStore.  I have a composite pattern in mind where we can manage all PersistentStores of the same type (GCEPD) in one PersistentStoreController, but internally there are still managed 1 per distinct PS.VolumeSource.